core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -835,22 +836,29 @@ private SubPlan updateStagesPartitioning(SubPlan plan)
                 }
 
                 // calculate (estimated) input data size to determine if we want to change number of partitions at runtime
-                long totalBytes = 0;
-                for (RemoteSourceNode remoteSourceNode : fragment.getRemoteSourceNodes()) {
-                    if (remoteSourceNode.getExchangeType() != REPLICATE) {
-                        for (PlanFragmentId sourceFragmentId : remoteSourceNode.getSourceFragmentIds()) {
-                            StageId sourceStageId = getStageId(sourceFragmentId);
-                            OutputDataSizeEstimate outputDataSizeEstimate = isReadyForExecutionResult.getSourceOutputSizeEstimates().get(sourceStageId);
-                            verify(outputDataSizeEstimate != null, "outputDataSizeEstimate not found for source stage %s", sourceStageId);
-                            totalBytes += outputDataSizeEstimate.getTotalSizeInBytes();
-                        }
-                    }
-                }
+                List<Long> partitionedInputBytes = fragment.getRemoteSourceNodes().stream()
+                        .filter(remoteSourceNode -> remoteSourceNode.getExchangeType() != REPLICATE)
+                        .map(remoteSourceNode -> remoteSourceNode.getSourceFragmentIds().stream()
+                                .mapToLong(sourceFragmentId -> {
+                                    StageId sourceStageId = getStageId(sourceFragmentId);
+                                    OutputDataSizeEstimate outputDataSizeEstimate = isReadyForExecutionResult.getSourceOutputSizeEstimates().get(sourceStageId);
+                                    verify(outputDataSizeEstimate != null, "outputDataSizeEstimate not found for source stage %s", sourceStageId);
+                                    return outputDataSizeEstimate.getTotalSizeInBytes();
+                                })
+                                .sum())
+                        .collect(toImmutableList());
+                // Currently the memory estimation is simplified:
+                // if it's an aggregation, then we use the total input bytes as the memory consumption
+                // if it's a multi-way join, we assume the smallest remote source will be streamed through
```

Comment:
we do not have multi-way joins - just 2-way.

Also "assume the smallest remote source will be streamed through"; are we assuming smalles will be streamed to be as conservative as possible? Can you clarify in comment?

Reply:
> we do not have multi-way joins - just 2-way.

Oh I meant multiple joins in the fragment.

> Also "assume the smallest remote source will be streamed through"; are we assuming smalles will be streamed to be as conservative as possible? Can you clarify in comment?

yeah it's meant to be conservative.